### PR TITLE
De-duplicate blogs after syncing

### PR DESCRIPTION
**Pull request by koke**
This doesn't prevent the site duplication from happening, but if it has happened it will clean it up in the next blog sync

To help with #7886

To test:

To set up the scene, check out `release/13.1` and run the app, but before set up a breakpoint in `-[BlogService updateBlogWithRemoteBlog:account:]` so it always creates a duplicate when syncing:

Screen Shot 2019-08-23 at 11 32 52

Then go to the sites list and pull to refresh until you start seeing duplicates.

Add some local drafts in different blogs by creating new posts and relaunching the app before they get saved on the server.

localDrafts

Now checkout `fix/deduplicate-blogs` and run the app. Go to the sites list and wait for a bit, it should sync automatically. If not, pull to refresh. The duplicates should be gone. If you go to the posts list, every local draft should be there.

localDraftsRecovered

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WordPress/Classes/Services/BlogService+Deduplicate.swift [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    /// If there's more than one blog in each group with local drafts, those will be reassigned
+    /// to the remaining blog.
+    ///
+    @objc(deduplicateBlogsForAccount:)
+    func deduplicateBlogs(for account: WPAccount) {
+        // Group all the account blogs by ID so it's easier to find duplicates
+        let blogsById = Dictionary(grouping: account.blogs, by: { $0.dotComID?.intValue ?? 0 })
+        // For any group with more than one blog, remove duplicates
+        for (blogID, group) in blogsById where group.count > 1 {
+            assert(blogID > 0, "There should not be a Blog without ID if it has an account")
```
Comment (jklausa) [inline WordPress/Classes/Services/BlogService+Deduplicate.swift]: Wouldn't this crash in cases where a XML-RPC blog would get duplicated? They will have a `dotComID` of nil, which would get coalesced to `0` in the line above and then trigger this `assert`.

Comment (jklausa) [reply WordPress/Classes/Services/BlogService+Deduplicate.swift]: Ah, but this is based on `Account` anyway, so they have to be `REST` blogs anyway. Nevermind.

File: WordPress/Classes/Services/BlogService+Deduplicate.swift [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    private func deduplicate(group: [Blog]) {
+        // If there's a blog with local drafts, we'll preserve that one, otherwise we pick up the first
+        // since we don't really care which blog to pick
+        let candidateIndex = group.firstIndex(where: { !localDrafts(for: $0).isEmpty }) ?? 0
+        let candidate = group[candidateIndex]
+
+        // We look through every other blog
+        for (index, blog) in group.enumerated() where index != candidateIndex {
```
Comment (jklausa) [inline WordPress/Classes/Services/BlogService+Deduplicate.swift]: I had no idea you could do a `where` together with `for... in`! That's so cool!

File: WordPress/Classes/Services/BlogService+Deduplicate.swift [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            DDLogInfo("Deleting duplicate blog \(blog.logDescription())")
+            managedObjectContext.delete(blog)
+        }
+    }
+
+    private func localDrafts(for blog: Blog) -> [AbstractPost] {
+        // The original predicate from PostService.countPostsWithoutRemote() was:
+        //   "postID = NULL OR postID <= 0"
+        // Swift optionals make things a bit more verbose, but this should be equivalent
+        return blog.posts?.filter({ (post) -> Bool in
```
Comment (jklausa) [inline WordPress/Classes/Services/BlogService+Deduplicate.swift]: There's a `isDraft` property on `AbstractPost` you might be able to use here, though it uses different logic — not sure which is more applicable.

File: WordPress/Classes/Services/BlogService.m [hunk 4]
```diff
@@ -671,6 +671,17 @@ - (void)mergeBlogs:(NSArray<RemoteBlog *> *)blogs withAccount:(WPAccount *)accou
...
+    /*
+     Sometimes bad things happen and blogs get duplicated. 👭
+     Hopefully we'll fix all the causes and this should never happen again 🤞🤞🤞
+     But even if it never happens again, it has already happened so we need to clean up. 🧹
+     Otherwise, users would have to reinstall the app to get rid of duplicates 🙅‍♀️
+
+     More context here:
+     https://github.com/wordpress-mobile/WordPress-iOS/issues/7886#issuecomment-524221031
+     */
+    [self deduplicateBlogsForAccount:account];
```
Comment (jklausa) [inline WordPress/Classes/Services/BlogService.m]: Should we do it on every `syncBlogsForAccount:` call? Seems like once during an app startup _should_ be enough?

File: WordPress/Classes/Services/BlogService+Deduplicate.swift (hunk 3 above)
Comment (koke) [reply WordPress/Classes/Services/BlogService+Deduplicate.swift]: I follow the same logic we use to show an alert warning that you have local drafts when signing out

File: WordPress/Classes/Services/BlogService.m (hunk 4 above)
Comment (koke) [reply WordPress/Classes/Services/BlogService.m]: We don't really know for sure how blogs get duplicated, and in any case I think it belongs conceptually in the sync method: when you call sync you expect the local list to reflect what's on the server.

The implementation will bail early if it detects no duplicates, so it shouldn't be too much overhead.

I wouldn't mind revisiting this once we have all the other fixes in place though.

**Pull request merged**